File: packages/core/src/shell/builtins/echo.ts
/**
 * echo - display a line of text
 */

import { BuiltinCommand, ShellEnvironment, ShellOptions } from '../types.js';

export class EchoCommand implements BuiltinCommand {
  name = 'echo';
  description = 'Display a line of text';

  async execute(args: string[], env: ShellEnvironment, options: ShellOptions): Promise<number> {
    let noNewline = false;
    let interpretEscapes = false;
    let startIndex = 0;

    // Parse options
    for (let i = 0; i < args.length; i++) {
      const arg = args[i];
      if (arg === '-n') {
        noNewline = true;
        startIndex = i + 1;
      } else if (arg === '-e') {
        interpretEscapes = true;
        startIndex = i + 1;
      } else if (arg === '-E') {
        interpretEscapes = false;
        startIndex = i + 1;
      } else if (arg.startsWith('-')) {
        // Unknown option, treat as text
        break;
      } else {
        break;
      }
    }

    // Join remaining arguments with spaces
    const text = args.slice(startIndex).join(' ');
    
    // Process escape sequences if -e is set
    const output = interpretEscapes ? this.processEscapes(text) : text;

    // Output the text
    if (noNewline) {
      process.stdout.write(output);
    } else {
      console.log(output);
    }

    return 0;
  }

  private processEscapes(text: string): string {
    return text
      .replace(/\\n/g, '\n')
      .replace(/\\t/g, '\t')
      .replace(/\\r/g, '\r')
      .replace(/\\b/g, '\b')
      .replace(/\\f/g, '\f')
      .replace(/\\v/g, '\v')
      .replace(/\\a/g, '\x07')
      .replace(/\\e/g, '\x1b')
      .replace(/\\\\/g, '\\')
      .replace(/\\0([0-7]{1,3})/g, (match, octal) => {
        return String.fromCharCode(parseInt(octal, 8));
      })
      .replace(/\\x([0-9a-fA-F]{1,2})/g, (match, hex) => {
        return String.fromCharCode(parseInt(hex, 16));
      });
  }

  help(): string {
    return `echo [-neE] [arg ...]
Display the args, separated by spaces, followed by a newline.

Options:
  -n    Do not append a newline
  -e    Enable interpretation of backslash escapes
  -E    Disable interpretation of backslash escapes (default)

Escape sequences (with -e):
  \\n    newline
  \\t    horizontal tab
  \\r    carriage return
  \\\\    backslash
  \\0nnn character with octal value nnn
  \\xhh  character with hex value hh`;
  }
}

File: shell/src/shell/builtins/echo.rs
//! echo - display line of text builtin command

use crate::shell::types::*;

pub struct EchoCommand;

#[async_trait::async_trait]
impl BuiltinCommand for EchoCommand {
    fn name(&self) -> &str {
        "echo"
    }

    fn description(&self) -> &str {
        "Display line of text"
    }

    async fn execute(
        &self,
        args: &[String],
        _env: &mut ShellEnvironment,
        _options: &ShellOptions,
    ) -> Result<i32, ShellError> {
        let mut newline = true;
        let mut start_index = 0;

        // Check for -n flag (no trailing newline)
        if !args.is_empty() && args[0] == "-n" {
            newline = false;
            start_index = 1;
        }

        // Print arguments separated by spaces
        if start_index < args.len() {
            print!("{}", args[start_index..].join(" "));
        }

        if newline {
            println!();
        }

        Ok(0)
    }

    fn help(&self) -> String {
        "echo [-n] [string ...]
Display the STRINGs on the standard output followed by a newline.

  -n    do not output the trailing newline".to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_echo_basic() {
        let mut env = ShellEnvironment::new();
        let echo = EchoCommand;
        
        let result = echo.execute(
            &["hello".to_string(), "world".to_string()], 
            &mut env, 
            &ShellOptions::default()
        ).await;
        
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }

    #[tokio::test]
    async fn test_echo_no_newline() {
        let mut env = ShellEnvironment::new();
        let echo = EchoCommand;
        
        let result = echo.execute(
            &["-n".to_string(), "hello".to_string()], 
            &mut env, 
            &ShellOptions::default()
        ).await;
        
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }

    #[tokio::test]
    async fn test_echo_empty() {
        let mut env = ShellEnvironment::new();
        let echo = EchoCommand;
        
        let result = echo.execute(&[], &mut env, &ShellOptions::default()).await;
        
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), 0);
    }
}